colexec: extract multiple new packages

yuzefovich opened the pull request:
This PR breaks down huge `colexec` package in order to speed up the build time
(it was pointed out that the package is the bottleneck during bazel builds). With
all of these changes, `colexec` (and its descendants) is barely visible when profiling
the bazel build.

The structure of dependencies is enforced by `dep_test` files added to the new
packages.

The following new packages have been extracted:
- `colexecutils` which contains miscellaneous utility operators, structs,
and functions that will be used by several other packages
- `colexechash` which contains all of the code interacting directly
with hashing of data (the hash table and the tuple hash distributor)
- `colexecwindow` which contains all of the code related to window
functions
- `colexecargs` which contains the arguments to and the result of
`NewColOperator` call
- `colexecbase` which contains miscellaneous operators that have very
few dependencies and are depended on by other operators that we want
to extract from `colexec`
- `colexeccmp` which contains shared objects between projection and
selection operators (namely, LIKE ops- and default comparison- related
things)
- `colexecproj` and `colexecsel` which contain the projection and the
selection operators, respectively
- `colexecjoin` which contains the code for the in-memory joiners (cross,
hash, and merge).

Additionally, multiple other operators have been moved to more appropriate
packages as well as the following package moves and renames were performed:
- move `colexecerror` out of `sql/colexecbase` into `sql`
- rename `sql/colexecbase` to `sql/colexecop`.

See individual commits for details.

yuzefovich committed:
colexec: extract new colexecutils package

This commit extracts new `colexecutils` package from `colexec`, and the
new package now contains the following:
- the code for the spilling queue (which has been exported)
- `AppendOnlyBufferedBatch` wrapper (also exported) and
`MakeWindowIntoBatch` utility method.

Additionally, this commit moves `FeedOperator` and `NonExplainable` marker
into the `colexecbase` package and extracts the utility method to generate
random batch size for tests from `colexec` into `colexectestutils`.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60770)
<!-- Reviewable:end -->

yuzefovich committed:
colexec: extract new colexechash package

This commit extracts new `colexechash` package that currently contains
all of the code for the hash table as well as some hash utilities (like
tuple hash distributor). In order to get here the following changes have
been made:
- many things about the hash table have been exported
- `TupleHashDistributor` has been exported
- `CancelChecker` and some utility methods have been moved to
`colexecutils` package.

The reason for extracting this package is that I envision that we will
have the operators that use the hash table (hash joiner, hash
aggregator, unordered distinct) live in separate packages, so having
separate "hash utility" package seems like a good idea.

Additionally, this commit begins enforcement of some dependencies
(mainly so that `colexecagg` and `colexechash` don't depend on each
other). The follow up commits will continue with this too. The goal is
making sure that we have enough packages that can be built in parallel.

Release note: None

yuzefovich committed:
colflow: move hash router into colflow package

This commit moves the hash router into `colflow` package since it is
a better place for it. This required some changes to the test in order
to instantiate the objects previously taken from `colexec/main_test.go`.
Additional complication is that `colflow/main_test.go` lives in
`colflow_test` package in order to avoid an import cycle, so we needed
to create some utility objects manually in `routers_test.go`.

Release note: None

yuzefovich committed:
colexec: break up operator.go file

This commits breaks up `operator.go` file in the following manner:
- move `OperatorInitStatus` into `colexecbase`
- export and move `CloserHelper` and `ClosableOperator` into
`colexecbase`
- move `twoInputNode` into newly created `colexec/joiner_utils.go` since
it is only used by two joiner operators
- export and move `OneInputCloserHelper` into `colexecbase`
- move `zeroOperator` and `fixedNumTuplesNoInputOp` into `colexecutils`
- move `vectorTypeEnforcer` and `BatchSchemaPrefixEnforcer` into
`colexecutils`.

Release note: None

yuzefovich committed:
colexec: extract new colexecwindow package

This commit extracts new `colexecwindow` package that currently contains
all of the code related to the window functions (several operators moved
from `colexec` as well as some utility information moved from
`colbuilder`). It is worth noting that at this moment this new package
depends on `colexec` for a few reasons (planning a Const operator,
a chain of ordered distinct operators, etc), but I intend to break up
that dependency later.

Release note: None

yuzefovich committed:
colexec: move a few operators to better places

This commit performs the following moves from `colexec` package:
- the deselector op into `colexecutils`
- the invariants checker and the panic injector into `colflow`
- the stats-related things are moved into `colflow` package and are
unexported; additionally, redundant `NetworkReader` interface is now
removed since it is no longer needed for mocking out the `colrpc.Inbox`.

Additionally, `defaultMemoryLimit` constant is moved into `colexecbase`
package.

Release note: None

yuzefovich committed:
colexec: extract op_creation into new colexecargs package

This commit extracts several structs that encompass the arguments to
and the result of `NewColOperator` call into new `colexecargs` package.

Additionally, `colexec/proj_utils_test.go` has been moved into
`colexectestutils` since in the future the utilities from the file will
be used from different packages.

Release note: None

yuzefovich committed:
colexec: break up distinct ops and sort partitioners

`distinct_tmpl` is a bit special because it is used to generate code for
two different purposes: distinct structs (used by the ordered distinct)
and partitioner structs (used by the sort operators). Previously, all of
this would get generated into the same file making it harder to break
down `colexec` package, so this commit updates the corresponding code
generator to write those two things into two separate files.

Release note: None

yuzefovich committed:
colexec: extract new colexecmisc package

This commit extracts new `colexecmisc` package that will contain
miscellaneous operators that don't have many dependencies, could be
grouped together, and don't have some "theme" around them.

(NOTE: eventually, this package will be named `colexecbase` and
`colexecmisc` is temporary name in order to avoid the name clash with
`sql/colexecbase` which will be renamed in a follow-up commit.)

The following things now live in `colexecmisc`:
- casts
- const operator
- ordered distinct
- ordinality operator
- simple project.

Additional changes:
- `BoolVecToSelOp` is exported and moved into `colexecutils`
- `ExternalSorterMinPartitions` constant is moved into `colexecbase`

All of this allowed to break the dependency of `colexecwindow` on
`colexec`.

Some changes to the bazel build of `colexec` were needed in order to
generate `colexec/sort_partitioner.eg.go` that depends on `distinct_tmpl`
now residing in `colexecmisc`.

Note that the unit tests for the ordered distinct are still kept in
`colexec` package since they are tightly coupled with the unordered and
external distinct unit tests.

Release note: None

yuzefovich committed:
colexec: extract projection and selection operators

This commit extracts new packages `colexecproj` and `colexecsel` to
contain the projection and the selection operators, respectively.
Additionally, a base `colexeccmp` package has been extracted which
contains the default comparison expression adapter as well as some
common code for LIKE operators.

Release note: None

yuzefovich committed:
colexec: extract new colexecjoin package

This commit moves the code for in-memory joiners (cross, hash, merge)
into new `colexecjoin` package. Most of the tests are still kept in
`colexec` since they are tightly coupled with the unit tests of the
external operators.

Release note: None

yuzefovich commented:
I think this is RFAL. The first commit should be ignored since it is a part of a separate PR.

rickystewart commented on pkg/sql/colexec/colexecmisc/COLEXECMISC.bzl:
```diff
@@ -0,0 +1,58 @@
+# NB: The helpers here were grafted from pkg/sql/colexec/COLEXEC.bzl.
```
I think this is too much code duplication. The helpers should be refactored into a common place. Let me know if you need help with this.

irfansharif reviewed (commented):
If there's already a common ("base") package that all these other packages depend on, the common bazel rules can be moved in there, and that would avoid the deduplication. If there isn't, we could create one (even if only for bazel). Ultimately the "bazel stuff" within colexec will be/is already owned by the sql exec team, so I'll defer to y'all for what you think is best. 

In any case, thanks for chopping all of this up, it's pretty great! Very excited to try it out.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60770#-:-MUAAR0z24TtsUp9Q1nt:b-49t729)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)



<!-- Sent from Reviewable.io -->

asubiotto commented:
> colexecmisc which contains miscellaneous operators that have very
few dependencies and are depended on by other operators that we want
to extract from colexec

can we put the colexecmisc stuff into colexecbase? Seems like both packages serve the same purpose.

> colexecprojsel which contains shared objects between projection and
selection operators (namely, LIKE ops- and default comparison- related
things)

😬  Don't think it's a good idea to have `colexecproj`, `colexecsel`, and `colexecprojsel`. Is there something better we can name it? Maybe `colexeccmp`?

yuzefovich commented:
> can we put the `colexecmisc` stuff into `colexecbase`? Seems like both packages serve the same purpose.

No, they serve different purposes. `colexecbase` defines the `Operator` interface whereas `colexecmisc` contains different things implementing that interface. Merging those two would be bad since it would introduce the dependency on the new stuff that `colexecmisc` brings into the users of `colexecbase` (which are pretty much all packages).

> Don't think it's a good idea to have `colexecproj`, `colexecsel`, and `colexecprojsel`. Is there something better we can name it? Maybe `colexeccmp`?

I disagree because both projection and selection operators result in huge amount of generated code, and they become the bottleneck. In fact, we might want to go even further - break down `colexecproj` into `colexecprojconst` and `colexecprojnonconst` or something like that, so that we can separate out huge projection_ops files (but this can wait).

asubiotto commented:
> No, they serve different purposes. colexecbase defines the Operator interface whereas colexecmisc contains different things implementing that interface. Merging those two would be bad since it would introduce the dependency on the new stuff that colexecmisc brings into the users of colexecbase (which are pretty much all packages).

Well then let's rename `colexecbase` to be `colexecop` and `colexecmisc` to be `colexecbase`. `misc` feels too vague.

> I disagree because both projection and selection operators result in huge amount of generated code, and they become the bottleneck. In fact, we might want to go even further - break down colexecproj into colexecprojconst and colexecprojnonconst or something like that, so that we can separate out huge projection_ops files (but this can wait).

I'm not talking about the number of packages. I think that's good. I specifically mean changing the `colexecprojsel` name.

yuzefovich commented:
Renaming `colexecbase` to `colexecop` sounds good to me. I think I will also extract `colexecerror` outside of `sql/colexecbase` folder and put it in `sql` folder.

I misunderstood your comment about `colexecprojsel`, renaming it sounds good.

yuzefovich commented:
First commit is #60993.

I removed the duplication of bazel-related things and renamed `colexecprojsel` to `colexeccmp`.

However, I'm encountering some issues when renaming/moving packages with `make bazel-generate` (first I tried to move `colexecerror` out of `sql/colexecbase` into `sql`, and then independently I tried renaming `colexecbase` to `colexecop`, and both times I'm getting some warnings locally, but the bazel build succeeds). I don't want to just rename `colexecmisc` to `colexecbase` because then we will have package name clash between `sql/colexecbase` and `sql/colexec/colexecbase`, so I think we should proceed with this PR as is and perform the renaming later.

asubiotto commented:
OK. To be clear, why will it be easier later? In any case, LGTM. We should probably merge this sooner rather than later to avoid having to rebase too much.

yuzefovich commented:
I'm not saying it will be easier to do the rename later, but that we should not wait for it to happen given that rebasing this PR is a pain. I don't understand why I'm seeing those warnings locally, so I'll need help from dev-infra team.

irfansharif commented:
> However, I'm encountering some issues when renaming/moving packages with make bazel-generate (first I tried to move colexecerror out of sql/colexecbase into sql, and then independently I tried renaming colexecbase to colexecop, and both times I'm getting some warnings locally, but the bazel build succeeds)

What are the warnings? It'd be helpful if you could show us how to repro it (is it a particular commit + some particular command)? I'm just going to rubber stamp this PR on Alfonso's behalf, reading his last comment.

yuzefovich committed:
colexecerror: move out of colexecbase directory

`colexecerror` is used by many packages and deserves to be top-level
folder in `sql` directory.

Release note: None

yuzefovich committed:
colexecbase: rename to colexecop

This commit renames `colexecbase` package to `colexecop` since latter
better reflect its purpose.

Release note: None

yuzefovich committed:
colexecmisc: rename to colexecbase

This commit renames `colexecmisc` package to `colexecbase` since the
former is too broad, and the latter was always the original intention
for the name, but it was blocked by (previously) similarly named package
in another directory.

Release note: None

yuzefovich commented:
Ok, I figured out the problem with warnings (as it turned out I had old path being imported in a couple of test files, and Goland for some reason couldn't find them when I was doing the search - probably busted caches). Anyway, I added 3 more commits to get to the currently desired state:
1. move `colexecerror` out of `sql/colexecbase` into `sql`
2. rename `sql/colexecbase` to `sql/colexecop`
3. rename `colexec/colexecmisc` to `colexec/colexecbase`.

I'll wait for #60993 to merge and will merge this afterwards.

Thanks for the feedback everyone!

yuzefovich commented:
bors r+

yuzefovich commented:
Oops, forgot the label.

bors r+
